src/linker/Linker.Steps/SealerStep.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public class SealerStep : BaseStep
+	{
+		HashSet<TypeDefinition> referencedBaseTypeCache;
+
+		public SealerStep ()
+		{
+		}
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			if (!Context.Optimizations.IsEnabled (CodeOptimizations.Sealer, assembly))
+				return;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		protected override void EndProcess ()
+		{
+			referencedBaseTypeCache = null;
+		}
+
+		bool IsSubclassed (TypeDefinition type)
+		{
+			if (referencedBaseTypeCache == null) {
+				referencedBaseTypeCache = new HashSet<TypeDefinition> ();
+				foreach (var a in Context.GetAssemblies ()) {
+					foreach (var s in a.MainModule.Types) {
+						var btd = s.BaseType?.Resolve ();
+						if (btd != null)
+							referencedBaseTypeCache.Add (btd);
+
+						if (s.HasNestedTypes) {
```

Comment:
It seems like this will not find types nested beyond the top level. This could be made recursive, or the `a.MainModule.Types` call could be changed to `a.MainModule.GetAllTypes()` extension method.